File: apps/windows/TeslaSync.App/shared-surfaces/Accordion.Model.cs
using TeslaSync.App.Core.Motion;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata for the <c>Accordion</c> shared surface — the native mirror of the module-level
/// constants in <c>web/src/components/ui/Accordion.tsx</c>. The web component is a collapsible disclosure
/// section: a header row (an optional leading icon, the title, an optional badge, optional trailing header
/// content) over an animated body reveal, driven by a controlled-or-uncontrolled open flag. This carries the
/// diagnostics slug, the token brush / corner keys the surface tints through (the web
/// <c>text-[var(--text-primary)]</c> title, <c>text-[var(--text-muted)]</c> icon, <c>border-white/[0.06]</c>
/// hairline and <c>rounded-xl</c> radius), the default header / body padding (web <c>px-4 py-3</c>), the body
/// reveal duration (web framer-motion <c>0.2</c>s) and the collapsed / expanded chevron rotation the native
/// disclosure chevron reproduces (web <c>ChevronDown</c> + <c>rotate-180</c>). UI-free so the mapping is
/// asserted in tests without a XAML runtime.
/// </summary>
public static class AccordionRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "Accordion";

    /// <summary>Token brush key for the title text (web <c>text-[var(--text-primary)]</c>).</summary>
    public const string TitleBrushKey = "TsColorTextPrimaryBrush";

    /// <summary>Token brush key for the leading icon (web <c>text-[var(--text-muted)]</c>).</summary>
    public const string IconBrushKey = "TsColorTextMutedBrush";

    /// <summary>Token brush key for the outer hairline + body divider (web <c>border-white/[0.06]</c> / <c>/[0.04]</c>).</summary>
    public const string BorderBrushKey = "TsColorBorderBrush";

    /// <summary>Corner-radius token key for the outer container (web <c>rounded-xl</c> ≈ 12&#160;px).</summary>
    public const string CornerRadiusKey = "TsRadiusMd";

    /// <summary>Body reveal duration in milliseconds at full motion (web framer-motion <c>transition={{ duration: 0.2 }}</c>).</summary>
    public const int BodyRevealDurationMs = 200;

    /// <summary>The chevron rotation when collapsed (web resting <c>ChevronDown</c>); the native disclosure chevron reproduces this.</summary>
    public const double ChevronCollapsedRotationDegrees = 0;

    /// <summary>The chevron rotation when expanded (web <c>open &amp;&amp; rotate-180</c>); the native disclosure chevron reproduces this.</summary>
    public const double ChevronExpandedRotationDegrees = 180;

    /// <summary>The default header / body padding (web <c>px-4 py-3</c> — 16&#160;px horizontal, 12&#160;px vertical).</summary>
    public static AccordionPadding DefaultContentPadding { get; } = AccordionPadding.Symmetric(horizontal: 16, vertical: 12);
}

/// <summary>
/// A four-sided padding value, the UI-free analogue of a WinUI <c>Thickness</c> — kept WinUI-free so the web
/// <c>px-4 py-3</c> defaults and any caller override (web <c>headerClassName</c> / <c>bodyClassName</c>) are
/// asserted in tests without a XAML runtime. The view maps it to a <c>Thickness</c> at the platform boundary.
/// </summary>
/// <param name="Left">The left inset in pixels.</param>
/// <param name="Top">The top inset in pixels.</param>
/// <param name="Right">The right inset in pixels.</param>
/// <param name="Bottom">The bottom inset in pixels.</param>
public readonly record struct AccordionPadding(double Left, double Top, double Right, double Bottom)
{
    /// <summary>A symmetric padding (web <c>px-{h} py-{v}</c>): the same horizontal inset left + right, the same vertical inset top + bottom.</summary>
    /// <param name="horizontal">The left + right inset in pixels.</param>
    /// <param name="vertical">The top + bottom inset in pixels.</param>
    public static AccordionPadding Symmetric(double horizontal, double vertical) =>
        new(horizontal, vertical, horizontal, vertical);

    /// <summary>A uniform padding — the same inset on every side.</summary>
    /// <param name="all">The inset applied to every side in pixels.</param>
    public static AccordionPadding Uniform(double all) => new(all, all, all, all);
}

/// <summary>
/// The render-time data model the <c>Accordion</c> view binds to — the native analogue of the web
/// <c>AccordionProps</c> (<c>Accordion.tsx</c> L6-27). The web component is purely presentational: its parent
/// owns any data fetching and feeds an already-resolved title, icon, badge and body, so — exactly like React
/// re-rendering the element with already-resolved props — there is no fetch-driven loading / error / stale /
/// offline branch to reproduce here; the only state is open vs. collapsed (carried by
/// <see cref="AccordionViewModel"/>) plus the optional icon / badge / header-extra composition. The arbitrary
/// React nodes (<c>icon</c> / <c>badge</c> / <c>headerExtra</c> / <c>children</c>) are surfaced as content slots
/// on the view; this model carries the WinUI-free data (the title, the optional leading glyph, whether a badge /
/// header-extra slot is filled, the initial open flag, the padding overrides and the test id) so the projection
/// is unit-tested without a UI host.
/// </summary>
public sealed record AccordionModel
{
    private AccordionModel(
        string title,
        string? iconGlyph,
        bool hasBadge,
        bool hasHeaderExtra,
        bool defaultOpen,
        AccordionPadding headerPadding,
        AccordionPadding bodyPadding,
        string? testId)
    {
        Title = title;
        IconGlyph = iconGlyph;
        HasBadge = hasBadge;
        HasHeaderExtra = hasHeaderExtra;
        DefaultOpen = defaultOpen;
        HeaderPadding = headerPadding;
        BodyPadding = bodyPadding;
        TestId = testId;
    }

    /// <summary>The header title text (web <c>title</c>).</summary>
    public string Title { get; }

    /// <summary>Optional leading Segoe Fluent glyph (web <c>icon</c>); null renders no icon — the web never auto-adds one.</summary>
    public string? IconGlyph { get; }

    /// <summary>Whether a trailing badge slot is filled (web <c>badge</c> present).</summary>
    public bool HasBadge { get; }

    /// <summary>Whether a trailing header-extra slot is filled (web <c>headerExtra</c> present, e.g. inline search).</summary>
    public bool HasHeaderExtra { get; }

    /// <summary>The initial open state for the uncontrolled mode (web <c>defaultOpen</c>, default <c>false</c>).</summary>
    public bool DefaultOpen { get; }

    /// <summary>Header padding (web <c>headerClassName</c> override; default <c>px-4 py-3</c>).</summary>
    public AccordionPadding HeaderPadding { get; }

    /// <summary>Body padding (web <c>bodyClassName</c> override; default <c>px-4 py-3</c>).</summary>
    public AccordionPadding BodyPadding { get; }

    /// <summary>Optional automation id mirroring a web test hook (→ <c>AutomationProperties.AutomationId</c>).</summary>
    public string? TestId { get; }

    /// <summary>An empty, collapsed accordion — the safe default the view falls back to when no model is supplied.</summary>
    public static AccordionModel Empty { get; } = Create(string.Empty);

    /// <summary>Build an accordion model.</summary>
    /// <param name="title">The header title text (web <c>title</c>).</param>
    /// <param name="iconGlyph">Optional leading Segoe Fluent glyph (web <c>icon</c>).</param>
    /// <param name="hasBadge">Whether a trailing badge slot is filled (web <c>badge</c>).</param>
    /// <param name="hasHeaderExtra">Whether a trailing header-extra slot is filled (web <c>headerExtra</c>).</param>
    /// <param name="defaultOpen">The initial open state for the uncontrolled mode (web <c>defaultOpen</c>).</param>
    /// <param name="headerPadding">Header padding override (web <c>headerClassName</c>); null uses the default.</param>
    /// <param name="bodyPadding">Body padding override (web <c>bodyClassName</c>); null uses the default.</param>
    /// <param name="testId">Optional automation id (web test hook).</param>
    public static AccordionModel Create(
        string title,
        string? iconGlyph = null,
        bool hasBadge = false,
        bool hasHeaderExtra = false,
        bool defaultOpen = false,
        AccordionPadding? headerPadding = null,
        AccordionPadding? bodyPadding = null,
        string? testId = null)
    {
        ArgumentNullException.ThrowIfNull(title);
        return new AccordionModel(
            title,
            iconGlyph,
            hasBadge,
            hasHeaderExtra,
            defaultOpen,
            headerPadding ?? AccordionRegistration.DefaultContentPadding,
            bodyPadding ?? AccordionRegistration.DefaultContentPadding,
            testId);
    }
}

/// <summary>
/// The fully projected, render-ready view of an <see cref="AccordionModel"/> at a given open state — everything
/// the web component derives before returning JSX (<c>Accordion.tsx</c> L50-83): the title and its
/// <see cref="TitleBrushKey"/>, the optional leading <see cref="IconGlyph"/> + its <see cref="IconBrushKey"/>,
/// whether the badge / header-extra slots render, the resolved <see cref="IsOpen"/> / <see cref="IsBodyVisible"/>
/// (web <c>open &amp;&amp; ...</c>), the <see cref="ChevronRotationDegrees"/> the disclosure chevron points to
/// (web <c>open &amp;&amp; rotate-180</c>), the header / body padding, the outer <see cref="BorderBrushKey"/> +
/// <see cref="CornerRadiusKey"/>, the composed <see cref="AutomationName"/> Narrator reads and the optional
/// <see cref="AutomationId"/>. Pure data so every value is asserted headlessly.
/// </summary>
/// <param name="Title">The header title text (web <c>title</c>).</param>
/// <param name="TitleBrushKey">Token brush key for the title (web <c>text-[var(--text-primary)]</c>).</param>
/// <param name="HasIcon">True when a leading icon is shown (web <c>icon &amp;&amp; ...</c>).</param>
/// <param name="IconGlyph">The optional leading glyph (web <c>icon</c>); null when absent.</param>
/// <param name="IconBrushKey">Token brush key for the icon (web <c>text-[var(--text-muted)]</c>).</param>
/// <param name="HasBadge">True when the trailing badge slot renders (web <c>badge</c>).</param>
/// <param name="HasHeaderExtra">True when the trailing header-extra slot renders (web <c>headerExtra</c>).</param>
/// <param name="IsOpen">The resolved open state.</param>
/// <param name="IsBodyVisible">True when the body is revealed (web <c>open &amp;&amp; &lt;motion.div&gt;</c>); equals <see cref="IsOpen"/>.</param>
/// <param name="ChevronRotationDegrees">The chevron rotation the disclosure points to (0 collapsed, 180 expanded — web <c>rotate-180</c>).</param>
/// <param name="HeaderPadding">The header padding (web <c>headerClassName</c> default <c>px-4 py-3</c>).</param>
/// <param name="BodyPadding">The body padding (web <c>bodyClassName</c> default <c>px-4 py-3</c>).</param>
/// <param name="BorderBrushKey">Token brush key for the outer hairline + body divider (web <c>border-white/[0.06]</c>).</param>
/// <param name="CornerRadiusKey">Corner-radius token key for the outer container (web <c>rounded-xl</c>).</param>
/// <param name="AutomationName">The accessible name Narrator reads for the disclosure (the title).</param>
/// <param name="AutomationId">The optional automation id (web test hook); null when not supplied.</param>
public sealed record AccordionDisplay(
    string Title,
    string TitleBrushKey,
    bool HasIcon,
    string? IconGlyph,
    string IconBrushKey,
    bool HasBadge,
    bool HasHeaderExtra,
    bool IsOpen,
    bool IsBodyVisible,
    double ChevronRotationDegrees,
    AccordionPadding HeaderPadding,
    AccordionPadding BodyPadding,
    string BorderBrushKey,
    string CornerRadiusKey,
    string AutomationName,
    string? AutomationId);

/// <summary>
/// Pure projection from an <see cref="AccordionModel"/> + the resolved open state to its
/// <see cref="AccordionDisplay"/> — the native port of <c>web/src/components/ui/Accordion.tsx</c>. Reproduces
/// the web derivations exactly: the title tints with the primary text token and the icon with the muted token;
/// the icon renders only when a glyph is supplied (web <c>icon &amp;&amp; ...</c>); the body is revealed and the
/// chevron points to 180° only while open (web <c>open &amp;&amp; &lt;motion.div&gt;</c> / <c>open &amp;&amp;
/// rotate-180</c>); the header / body padding pass through (web <c>headerClassName</c> / <c>bodyClassName</c>);
/// and the accessible name is the title (the natural Narrator label for the disclosure). No WinUI types — so the
/// projection is unit-tested without a UI host.
/// </summary>
public static class AccordionProjection
{
    /// <summary>Leading icon font size — web <c>h-4 w-4</c> (≈16&#160;px).</summary>
    public const double IconSize = 16;

    /// <summary>Title font size — web <c>text-sm</c> (0.875rem ≈ 14&#160;px).</summary>
    public const double TitleFontSize = 14;

    /// <summary>Spacing between the header items — web <c>gap-3</c> (≈12&#160;px).</summary>
    public const double HeaderItemSpacing = 12;

    /// <summary>Project <paramref name="model"/> at <paramref name="isOpen"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web props).</param>
    /// <param name="isOpen">The resolved open state (from <see cref="AccordionViewModel"/>).</param>
    /// <param name="localizer">The i18n facade (P1/S10); reserved for parity with the surface family — this anonymous surface carries no inherent strings, so the title / badge / header-extra content is caller-supplied and already localized.</param>
    /// <returns>The render-ready display model.</returns>
    public static AccordionDisplay Project(AccordionModel model, bool isOpen, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        bool hasIcon = !string.IsNullOrEmpty(model.IconGlyph);

        return new AccordionDisplay(
            Title: model.Title,
            TitleBrushKey: AccordionRegistration.TitleBrushKey,
            HasIcon: hasIcon,
            IconGlyph: hasIcon ? model.IconGlyph : null,
            IconBrushKey: AccordionRegistration.IconBrushKey,
            HasBadge: model.HasBadge,
            HasHeaderExtra: model.HasHeaderExtra,
            IsOpen: isOpen,
            IsBodyVisible: isOpen,
            ChevronRotationDegrees: isOpen
                ? AccordionRegistration.ChevronExpandedRotationDegrees
                : AccordionRegistration.ChevronCollapsedRotationDegrees,
            HeaderPadding: model.HeaderPadding,
            BodyPadding: model.BodyPadding,
            BorderBrushKey: AccordionRegistration.BorderBrushKey,
            CornerRadiusKey: AccordionRegistration.CornerRadiusKey,
            AutomationName: model.Title,
            AutomationId: model.TestId);
    }
}

/// <summary>
/// Pure reduce-motion gating for the body reveal — the native analogue of the web framer-motion entrance
/// (<c>Accordion.tsx</c> L67-79: <c>opacity 0 → 1</c> over <c>0.2</c>s). The native disclosure animates its own
/// height + chevron and already honours the OS "show animations" setting at the platform layer; this gates the
/// surface's additional body fade so that, under reduced motion, the revealed body appears immediately in its
/// final state (the duration collapses to 0&#160;ms, mirroring framer-motion's <c>initial={false}</c>). Kept
/// static + WinUI-free so the policy is unit-tested without a <c>UISettings</c> host.
/// </summary>
public static class AccordionMotion
{
    /// <summary>
    /// The effective body-reveal fade duration in milliseconds: <see cref="AccordionRegistration.BodyRevealDurationMs"/>
    /// at full motion, 0 under reduced motion.
    /// </summary>
    /// <param name="reduceMotion">Whether the OS reduce-motion preference is set (web <c>prefers-reduced-motion</c>).</param>
    public static int BodyRevealDurationMs(bool reduceMotion) =>
        MotionDuration.Resolve(reduceMotion, AccordionRegistration.BodyRevealDurationMs);

    /// <summary>Whether the body fade should run at all — false under reduced motion (render the final state immediately).</summary>
    /// <param name="reduceMotion">Whether the OS reduce-motion preference is set (web <c>prefers-reduced-motion</c>).</param>
    public static bool ShouldAnimateBody(bool reduceMotion) => MotionDuration.ShouldAnimate(reduceMotion);
}

/// <summary>
/// PII-safe diagnostics for the <c>Accordion</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never the title or body content — so a
/// diagnostics line can never leak fleet state. Thread-safe; mirrors the peer surfaces' diagnostics collectors.
/// </summary>
public sealed class AccordionDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">An optional sink the operational event line is written to.</param>
    public AccordionDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=Accordion</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={AccordionRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/Accordion/AccordionModel.kt
// Pure, framework-free model + render classifier + a11y affordances + diagnostics for the Accordion shared
// surface — the native analogue of every decision the web component makes
// (web/src/components/ui/Accordion.tsx) before it paints. No Compose, no Android, no HTTP: every declaration
// here is unit-tested off-device in the :app:testReleaseUnitTest gate, keeping the composable a thin render
// layer.
//
// What the web source is (and therefore the COMPLETE branch set this surface reproduces):
//   • A PURE, presentational component — "Collapsible content section with animated reveal" (web JSDoc): a
//     rounded, hairline-bordered container whose header is a single `<button>` carrying an optional leading
//     `icon`, the `title`, an optional `badge`, an optional `headerExtra` slot, and a trailing `ChevronDown`
//     that rotates 180° while open; below it an `AnimatePresence` reveals the `children` body (a top divider
//     + padded content) when open. The parent owns ALL content (title / icon / badge / headerExtra / body);
//     the component's only logic is owning (or, when controlled, deferring) the open/closed boolean and
//     selecting which optional regions to draw. There is NO hook, NO fetch, and NO data port to bind (no
//     P1/S8 state holder, no Source/ViewModel) — modelling one would invent an async dependency the web spec
//     does not have (honesty covenant: no scope narrowing, no silent drift). The sibling presentational ports
//     InlineCallout / ProgressRing / ReleaseNotes set the same precedent (composable + pure model, no
//     Source/ViewModel).
//   • So the surface's REAL, fully-reproduced states are the open/closed toggle crossed with its prop-driven
//     branches: a leading icon present/absent (web `{icon && …}`), a badge present/absent (web `{badge}`), a
//     headerExtra present/absent (web `{headerExtra}`), and a real body vs an empty body (web `children`).
//     Each is reduced here in [classifyAccordion] and asserted in the off-device test, doubling as the
//     per-state snapshot.
//   • Controlled vs uncontrolled is the one genuinely stateful decision the web makes
//     (`isControlled = open !== undefined && onOpenChange !== undefined`; `open = isControlled ? open :
//     internalOpen`). It is reproduced verbatim by [accordionIsControlled] / [resolveAccordionOpen] so the
//     composable's `remember`-backed fallback and the parent-owned override behave exactly like the source.
//   • The one place this surface improves on a literal port is the empty-body branch: the web renders whatever
//     `children` it is given (including nothing), but the prompt's "empty → friendly empty state, never a
//     blank box" contract is honoured by [classifyAccordion] flagging an empty body so the view renders a
//     localized caption instead of an empty region.
//
// Why the generic data-surface states (loading / error / stale / offline) are intentionally absent: this
// surface fetches nothing — it is a controlled container whose content is handed in by its parent. There is no
// query to be loading, to fail, to go stale, or to be offline, so inventing those states would be dishonest.
// The owning screen that DOES fetch (and can be loading/stale/offline) renders its own data surface and only
// mounts this accordion once it already has content to reveal.
//
// Accessibility: the web header is a `<button aria-expanded={open}>`. A DOM button announces its inner text
// and its expanded state for free; Compose has no auto-localized `aria-expanded`, so the native header carries
// a localized action label (Expand / Collapse) and state description (Expanded / Collapsed). The web source
// owns no text keys for these (it relies on the DOM), so — exactly like the sibling ReleaseNotes surface —
// they resolve by-name through the i18n facade ([resolveOptional], the native mirror of i18next `t(key,
// default)`) with the English [AccordionDefaults] fallbacks. The only other string the surface shows beyond
// its caller-supplied props is the empty-body caption, which resolves through the existing
// `translation_common_noData` catalog key at the render boundary (P1/S10).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/Accordion — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName`
// is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.accordion

import io.teslasync.shared.core.diagnostics.Logger

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no title, badge, or body —
 * only this constant identifier — so a diagnostics line can never leak the accordion's content.
 */
const val ACCORDION_SLUG: String = "Accordion"

/**
 * Whether the accordion is in controlled mode — the native mirror of the web
 * `isControlled = open !== undefined && onOpenChange !== undefined`. Controlled mode means the parent owns the
 * open/closed source of truth; both an [openOverride] and an [hasOnOpenChange] handler must be supplied,
 * matching the web's requirement that the pair travel together (an `open` with no `onOpenChange` would be a
 * read-only prop the user could never toggle, so the web — and this port — fall back to internal state).
 */
fun accordionIsControlled(
    openOverride: Boolean?,
    hasOnOpenChange: Boolean,
): Boolean = openOverride != null && hasOnOpenChange

/**
 * Resolve the effective open state — web `open = isControlled ? openProp : internalOpen`. In controlled mode
 * the parent-owned [openOverride] wins; otherwise the composable's own remembered [internalOpen] is used. Pure
 * so the controlled/uncontrolled decision is unit-tested without a Compose host. A controlled call always
 * supplies a non-null [openOverride] (guaranteed by [accordionIsControlled]); the `?: internalOpen` is a
 * defensive fall-through that can never be reached in controlled mode.
 */
fun resolveAccordionOpen(
    openOverride: Boolean?,
    hasOnOpenChange: Boolean,
    internalOpen: Boolean,
): Boolean = if (accordionIsControlled(openOverride, hasOnOpenChange)) openOverride ?: internalOpen else internalOpen

/**
 * The parent-owned inputs to the accordion, bundled into one value object so the pure [classifyAccordion]
 * reads a single argument instead of a long parameter list — the native mirror of the web `AccordionProps`
 * region flags the parent supplies.
 *
 * @property expanded whether the body is currently revealed (the resolved web `open`).
 * @property hasIcon whether the parent supplied a leading glyph slot (web truthy `icon`).
 * @property hasBadge whether the parent supplied a badge slot (web truthy `badge`).
 * @property hasHeaderExtra whether the parent supplied a header-extra slot (web truthy `headerExtra`).
 * @property hasBody whether the parent supplied a non-empty body slot (web non-empty `children`).
 */
data class AccordionInput(
    val expanded: Boolean,
    val hasIcon: Boolean = false,
    val hasBadge: Boolean = false,
    val hasHeaderExtra: Boolean = false,
    val hasBody: Boolean = false,
)

/**
 * The render-ready classification of the accordion — everything the view needs to draw, reduced from the
 * parent's props so every branch is exhaustively covered and unit-tested off-device. The web component always
 * renders the container + header (the parent decides whether to mount it), so there is no hidden surface —
 * only which header regions are shown, whether the body is revealed, and whether the revealed body is real or
 * the empty fallback.
 *
 * @property expanded the body is revealed (web `open`); drives the chevron rotation and the body visibility.
 * @property showIcon a leading glyph is shown — only when the parent supplied one (web `{icon && …}`).
 * @property showBadge a badge is shown (web `{badge}`).
 * @property showHeaderExtra a header-extra slot is shown (web `{headerExtra}`).
 * @property showBody a real body is shown when expanded (web non-empty `children`).
 * @property showEmptyFallback no body was supplied — the view shows a localized caption when expanded, never a
 *   blank box.
 */
data class AccordionRender(
    val expanded: Boolean,
    val showIcon: Boolean,
    val showBadge: Boolean,
    val showHeaderExtra: Boolean,
    val showBody: Boolean,
    val showEmptyFallback: Boolean,
)

/**
 * Reduce the parent's [input] into the render-ready [AccordionRender]. Pure (no Compose). A body that is
 * present is shown; an absent body sets [AccordionRender.showEmptyFallback] so the revealed region is never a
 * blank box (the prompt's empty-state contract). The empty/real distinction is independent of [expanded] so
 * the classification is stable across the toggle and the test can assert each region once.
 */
fun classifyAccordion(input: AccordionInput): AccordionRender =
    AccordionRender(
        expanded = input.expanded,
        showIcon = input.hasIcon,
        showBadge = input.hasBadge,
        showHeaderExtra = input.hasHeaderExtra,
        showBody = input.hasBody,
        showEmptyFallback = !input.hasBody,
    )

/**
 * Native-only accessibility microcopy defaults — the web header is a `<button aria-expanded>` whose state +
 * action a DOM screen reader announces for free, so the web source owns no catalog key for these. Absent a
 * catalog hit these English fallbacks are used (the native mirror of i18next's default argument).
 */
object AccordionDefaults {
    /** Action label announced for the header while collapsed (TalkBack "double-tap to …"). */
    const val EXPAND_ACTION: String = "Expand"

    /** Action label announced for the header while expanded. */
    const val COLLAPSE_ACTION: String = "Collapse"

    /** State description announced while open — native equivalent of `aria-expanded="true"`. */
    const val EXPANDED_STATE: String = "Expanded"

    /** State description announced while closed — native equivalent of `aria-expanded="false"`. */
    const val COLLAPSED_STATE: String = "Collapsed"
}

/** Resource name for the header's expand action label (by-name; absent ⇒ the English default). */
const val KEY_ACCORDION_EXPAND_ACTION: String = "translation_accordion_expand"

/** Resource name for the header's collapse action label (by-name; absent ⇒ default). */
const val KEY_ACCORDION_COLLAPSE_ACTION: String = "translation_accordion_collapse"

/** Resource name for the header's expanded state description (by-name; absent ⇒ default). */
const val KEY_ACCORDION_EXPANDED_STATE: String = "translation_accordion_expanded"

/** Resource name for the header's collapsed state description (by-name; absent ⇒ default). */
const val KEY_ACCORDION_COLLAPSED_STATE: String = "translation_accordion_collapsed"

/**
 * The localized accessibility affordance strings for the collapsible header — resolved once at the render
 * boundary so the header's TalkBack action + state description track the open/closed toggle the way the web
 * `aria-expanded` does. The web source owns no text keys for these (it relies on the DOM), so they resolve
 * by-name with the English fallbacks above.
 */
data class AccordionAffordances(
    val expandAction: String,
    val collapseAction: String,
    val expandedState: String,
    val collapsedState: String,
) {
    /** The TalkBack action label for the toggle in its current [expanded] state. */
    fun actionLabel(expanded: Boolean): String = if (expanded) collapseAction else expandAction

    /** The TalkBack state description for the current [expanded] state (web `aria-expanded`). */
    fun stateLabel(expanded: Boolean): String = if (expanded) expandedState else collapsedState
}

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns the [lookup] result for
 * [resourceName] when it resolves to a non-blank string, otherwise the [fallback] default. [lookup] is a thin
 * seam over the Android string catalog in production (an optional by-name resource read) and a map in tests,
 * so the resolve-or-fallback decision stays pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the title,
 * badge, or body — so a diagnostics line can never leak what the accordion shows.
 */
object AccordionDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = ACCORDION_SLUG

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
